src/java/org/apache/cassandra/service/accord/async/AsyncLoader.java

```diff
@@ -0,0 +1,232 @@
+
+            futures.add(future);
+        }
+
+        return futures;
+    }
+
+    @VisibleForTesting
+    Function<AccordCommand, Future<?>> loadCommandFunction(Object callback)
+    {
+        return command -> Stage.READ.submit(() -> {
+            try
+            {
+                logger.trace("Starting load of {} for {}", command.txnId(), callback);
+                AccordKeyspace.loadCommand(command);
+                logger.trace("Completed load of {} for {}", command.txnId(), callback);
+            }
+            catch (Throwable t)
+            {
+                logger.error(String.format("Exception loading %s for %s", command.txnId(), callback), t);
+                throw t;
+            }
+        });
+    }
+
+    @VisibleForTesting
+    Function<AccordCommandsForKey, Future<?>> loadCommandsPerKeyFunction(Object callback)
+    {
+        return cfk -> Stage.READ.submit(() -> {
+            try
+            {
+                logger.trace("Starting load of {} for {}", cfk.key(), callback);
+                AccordKeyspace.loadCommandsForKey(cfk);
+                logger.trace("Completed load of {} for {}", cfk.key(), callback);
+            }
+            catch (Throwable t)
+            {
+                logger.error(String.format("Exception loading %s for %s", cfk.key(), callback), t);
+                throw t;
+            }
+        });
+    }
+
+    private Future<?> referenceAndDispatchReads(AsyncContext context, Object callback)
+    {
+        List<Future<?>> futures = null;
+
+        futures = referenceAndDispatchReads(txnIds,
+                                            commandStore.commandCache(),
+                                            context.commands.items,
+                                            loadCommandFunction(callback),
+                                            futures,
+                                            callback);
+
+        futures = referenceAndDispatchReads(keys,
+                                            commandStore.commandsForKeyCache(),
+                                            context.commandsForKey.items,
+                                            loadCommandsPerKeyFunction(callback),
+                                            futures,
+                                            callback);
+
+        return futures != null ? FutureCombiner.allOf(futures) : null;
+    }
+
+    public boolean load(AsyncContext context, BiConsumer<Object, Throwable> callback)
+    {
+        logger.trace("Running load for {} with state {}: {} {}", callback, state, txnIds, keys);
+        commandStore.checkInStoreThread();
+        switch (state)
+        {
+            case INITIALIZED:
+                state = State.SETUP;
+            case SETUP:
+                // notify any pending write only groups we're loading a full instance so the pending changes aren't removed
+                txnIds.forEach(commandStore.commandCache()::lockWriteOnlyGroupIfExists);
```

Comment:
Do we need some kind of rollback logic in the event of an unexpected exception?

Reply:
no, we can't unlock in that case, however the cache will remove the group on eviction, so it doesn't stay there forever